### ISSUE TITLE
Questions about averaging checkpoints in ASR

### ISSUE THREAD

**Issue by kehanlu** [espnet/espnet#4734]
Hi, 
averaging checkpoints is a common procedure in Espnet to obtain better final performance, but I wonder if there are literatures to explain why this technique useful? Are there any hidden behavior that impact the performance gain?

In my case, I am recently training the Transformer+CTC ASR model on TEDlium2 dataset. When I decode the invdividual epoch checkpoints the WERs are consistantly improved over baseline model. But when I average the model checkpoints(best/last few epochs), the performance gain are not as expected.

For example,
| decode(epoch) | 46  | 47  | 48  | 49  | 50  | avg 5 models|
|---------------|-----|-----|-----|-----|-----|------------|
| baseline(WER)   | 8.1 | 8.0 | 7.7 | 7.9 | 8.0 | 7.0        |
| baseline+aux loss(WER)   | 7.5 | 7.4 | 7.5 | 7.3 | 7.6 | 7.0        |

To be specific, I use bert wordpiece to tokenize the target sequence for both models. There are 20K tokens for TEDlium2 dataset, much more than the default recipe(500 BPE). 

Does anyone have experience about this phenomenon?

**Comment by sw005320** [comment]
Thanks for your discussion!
This is a very interesting result.
I usually only check the average checkpoint result, and I do not notice such behavior.

@ShigekiKarita, I think you mentioned some literature about it.
Can you answer it?

**Comment by ShigekiKarita** [comment]
Interesting. I'm afraid that model averaging is super adhoc during selecting best checkpoints. Maybe it is better if you select best ones by the lowest baseline+aux loss by tweaking the metric here https://github.com/espnet/espnet/blob/master/utils/average_checkpoints.py#L20

Personally, I prefer "last few epochs" or [exponential moving average](https://discuss.pytorch.org/t/how-to-apply-exponential-moving-average-decay-for-variables/10856/11) because I don't wanna struggle with the model selection, and save a bunch of checkpoints.

**Comment by kehanlu** [comment]
@sw005320  @ShigekiKarita Thanks for your reply, 
I actually tried averaging the last or best(valid CER, WER, loss ctc) few(5, 10) epochs. The selection of the metrics do change the performance, but the trend is similar. No matter what metric I choose, when I apply averaging checkpoint technique, the performance gain from individual epochs vanished.

Do you think this is because the models are not fully converge yet? The distribution in the last/best checkpoints may not be stable enough.

**Comment by ShigekiKarita** [comment]
Yes, it can. When I reported this Transformer result, it was improving at 100th epoch as seen its config.
> e2e JSON file: ./exp/train_trim_sp_pytorch_nbpe500_ngpu2_train_pytorch_transformer.v2_epochs100_specaug/results/model.json

https://github.com/espnet/espnet/blob/master/egs/tedlium2/asr1/RESULTS.md#transformer-large-model--specaug--large-lm
